### ISSUE TITLE
Backend connection: circuit breaker seems to loop indefinitively

### ISSUE THREAD

**Issue by BlackYoup** [sozu-proxy/sozu#260]
I had a few of those occurrences when I had an application with only one backend which refused connections:
```
2017-10-20T09:06:32Z    491102912933605 4477    INFO    TAG-0   Connecting xxx -> (V4(127.0.0.1:8081), 0, 331475)                                                                            
2017-10-20T09:06:32Z    491102912964060 4478    INFO    TAG-1   Connecting xxx -> (V4(127.0.0.1:8081), 0, 239414)                                                                            
2017-10-20T09:06:32Z    491102912993311 4477    ERROR   TAG-0   backend (V4(127.0.0.1:8081), 1) connections failed 331475 times, disabling it                                                     
2017-10-20T09:06:32Z    491102913033918 4478    ERROR   TAG-1   backend (V4(127.0.0.1:8081), 1) connections failed 239414 times, disabling it                                                     
2017-10-20T09:06:32Z    491102913103214 4478    ERROR   TAG-1   error connecting to backend, trying again                                                                                         
2017-10-20T09:06:32Z    491102913156742 4477    ERROR   TAG-0   error connecting to backend, trying again
```

It just keeps showing those messages, the try count increments without a 503 sent to the client

**Comment by mbStavola** [comment]
Seems like there should be a call to either the Backend's `can_open()` or the RetryPolicy's `can_try()` in `try_connect()`. The policy is probably in a failed state but it never conveys that information to the backend outside of `can_open()`.

**Comment by mbStavola** [comment]
For testing this, I just had a simple configuration with two apps: one pointing to nothing and the other running a [fake backend](https://github.com/micromata/http-fake-backend). I played with the app config using the config file watcher to watch changes. Not sure if there was a better way to go about this?

Anyway, I see that `TcpStream::connect()` always returns as `Ok` since, I assume, the connection _is_ technically okay... just still trying to connect. This means that it never has a chance to utilize circuit breaking logic since it's considered healthy. I guess this is related to the TODO just above `TcpStream::connect()` (re: EINPROGESS)? For some strange reason, this only happens when sticky sessions are turned on as non-sticky routing just fails with "cannot write default answer, back buffer was empty."

Just curious, is there a way to set a timeout for a backend in a sozu config?

**Comment by BlackYoup** [comment]
There was a way to set timeouts, as you can see on some commented code, but it has been disabled for now, see https://github.com/sozu-proxy/sozu/pull/166#issuecomment-296593522

My setup about this bug was an frontend, without sticky sessions and only one backend which wasn't available (connection refused error). I loaded the page using a browser (chrome here) and sozu kept trying to connect without sending a response to the client, which made the page load indefinitely.

I can dig into this issue a bit more but I don't think I will be able to do it before a few days :)

**Comment by BlackYoup** [comment]
Alright, I saw the same thing as you, I didn't thought the `connect()` would be non-blocking even if it makes sense.

But I confirm I also reproduce this without sticky sessions.

**Comment by Geal** [comment]
Hello, this should be fixed with cd9fce8

**Comment by BlackYoup** [comment]
It is indeed
